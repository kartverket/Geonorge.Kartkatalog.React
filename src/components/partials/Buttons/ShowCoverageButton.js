import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import style from './Buttons.scss';
import Modal from './Modal.js'

import { getResource } from '../../../actions/ResourceActions'

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export class ShowCoverageButton extends Component {
    constructor(props) {
        super(props);
        this.state = {
            show: false,
            mounted: false
        };
    }

    showModal = () => {
        this.setState({ show: true });
    };

    hideModal = () => {
        this.setState({ show: false });
    };

    componentDidMount = () => {
        this.setState({ mounted: true });
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (prevProps.selectedLanguage !== this.props.selectedLanguage){
            this.render();
        }
    }

    renderModal() {
        return this.props.metadata.CoverageUrl && this.state.mounted
            ? (
                <Modal show={this.state.show} handleClose={this.hideModal}>
                    <iframe src={this.state.show ? this.props.metadata.CoverageUrl : ''} width="100%" height="800px" />
                </Modal>
            ) : '';
    }

    renderButton() {
        let buttonDescription = this.props.getResource('DisplayCoverageMap', 'Vis dekningskart');

        if (this.props.metadata.CoverageUrl) {
            let buttonClass = style.btn;
            return <span className={buttonClass} onClick={this.showModal}>
                <FontAwesomeIcon title={buttonDescription} icon={['far', 'globe']} key="icon" />{buttonDescription}
            </span>
        }
        else {
            let buttonClass = style.btn + ' disabled';
            return <span className={buttonClass}>
                <FontAwesomeIcon title={buttonDescription} icon={['far', 'globe']} key="icon" />{buttonDescription}
            </span>
        }

    }

    render() {
        return (
            <React.Fragment>
                {this.renderModal()}
                {this.renderButton()}
            </React.Fragment>
        );
    }

}

ShowCoverageButton.propTypes = {
    metadata: PropTypes.object.isRequired
};

const mapDispatchToProps = {
    getResource
};

const mapStateToProps = state => ({
    selectedLanguage: state.selectedLanguage
});

export default connect(mapStateToProps, mapDispatchToProps)(ShowCoverageButton);
