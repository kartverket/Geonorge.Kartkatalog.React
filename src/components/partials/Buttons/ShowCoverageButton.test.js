import React from 'react';
import { shallow } from "enzyme";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { ShowCoverageButton } from './ShowCoverageButton';
import style from './Buttons.scss'

function setupItemWithCoverageUrl() {
    const metadata = {
        Uuid: '1234',
        CoverageUrl: 'test'
    }
    const props = {
        metadata: metadata
    }
    const wrapper = shallow(<ShowCoverageButton {...props} />)

    return {
        props,
        wrapper
    }
}

function setupItemWithoutCoverageUrl() {
    const metadata = {
        Uuid: '1234',
    }
    const props = {
        metadata: metadata
    }
    const wrapper = shallow(<ShowCoverageButton {...props} />)

    return {
        props,
        wrapper
    }
}


describe('setupItemWithCoverageUrl', () => {
    
    it('should render self', () => {
        const { wrapper } = setupItemWithCoverageUrl()
        expect(wrapper).toMatchSnapshot();
    })

    it('Item with coverage url', () => {
        const { wrapper } = setupItemWithCoverageUrl()

        wrapper.setState({
            show: true
        });
        
        expect(wrapper.children('span').hasClass(style.btn)).toBe(true)
        expect(wrapper.children('span').prop("onClick").toString()).toContain("show: true")
        expect(wrapper.find(FontAwesomeIcon).first().prop("icon")).toContain('globe')        
        expect(wrapper.find("span").first().html()).toContain('Vis dekningskart')
        expect(wrapper.find("iframe").prop("src")).toBe("test")
    })

    it('Item without coverage url', () => {
        const { wrapper } = setupItemWithoutCoverageUrl()
        
        expect(wrapper.children('span').hasClass(style.btn + ' disabled')).toBe(true)
        expect(wrapper.find(FontAwesomeIcon).first().prop("icon")).toContain('globe')        
        expect(wrapper.find("span").first().html()).toContain('Vis dekningskart')
    })
})
