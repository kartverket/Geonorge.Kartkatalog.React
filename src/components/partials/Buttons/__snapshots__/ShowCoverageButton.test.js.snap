// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupItemWithCoverageUrl should render self 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShowCoverageButton
    metadata={
      Object {
        "CoverageUrl": "test",
        "Uuid": "1234",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Connect(Modal)
          handleClose={[Function]}
          show={false}
        >
          <iframe
            height="800px"
            src="test"
            width="100%"
          />
        </Connect(Modal)>,
        <span
          className="btn"
          onClick={[Function]}
        >
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon={
              Array [
                "far",
                "globe",
              ]
            }
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title="Vis dekningskart"
            transform={null}
          />
          Vis dekningskart
        </span>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <iframe
            height="800px"
            src="test"
            width="100%"
          />,
          "handleClose": [Function],
          "show": false,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "height": "800px",
            "src": "test",
            "width": "100%",
          },
          "ref": null,
          "rendered": null,
          "type": "iframe",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Array [
                  "far",
                  "globe",
                ]
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              symbol={false}
              title="Vis dekningskart"
              transform={null}
            />,
            "Vis dekningskart",
          ],
          "className": "btn",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "icon",
            "nodeType": "function",
            "props": Object {
              "border": false,
              "className": "",
              "fixedWidth": false,
              "flip": null,
              "icon": Array [
                "far",
                "globe",
              ],
              "inverse": false,
              "listItem": false,
              "mask": null,
              "pull": null,
              "pulse": false,
              "rotation": null,
              "size": null,
              "spin": false,
              "symbol": false,
              "title": "Vis dekningskart",
              "transform": null,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "Vis dekningskart",
        ],
        "type": "span",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Connect(Modal)
            handleClose={[Function]}
            show={false}
          >
            <iframe
              height="800px"
              src="test"
              width="100%"
            />
          </Connect(Modal)>,
          <span
            className="btn"
            onClick={[Function]}
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Array [
                  "far",
                  "globe",
                ]
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              symbol={false}
              title="Vis dekningskart"
              transform={null}
            />
            Vis dekningskart
          </span>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <iframe
              height="800px"
              src="test"
              width="100%"
            />,
            "handleClose": [Function],
            "show": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "height": "800px",
              "src": "test",
              "width": "100%",
            },
            "ref": null,
            "rendered": null,
            "type": "iframe",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <FontAwesomeIcon
                border={false}
                className=""
                fixedWidth={false}
                flip={null}
                icon={
                  Array [
                    "far",
                    "globe",
                  ]
                }
                inverse={false}
                listItem={false}
                mask={null}
                pull={null}
                pulse={false}
                rotation={null}
                size={null}
                spin={false}
                symbol={false}
                title="Vis dekningskart"
                transform={null}
              />,
              "Vis dekningskart",
            ],
            "className": "btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "icon",
              "nodeType": "function",
              "props": Object {
                "border": false,
                "className": "",
                "fixedWidth": false,
                "flip": null,
                "icon": Array [
                  "far",
                  "globe",
                ],
                "inverse": false,
                "listItem": false,
                "mask": null,
                "pull": null,
                "pulse": false,
                "rotation": null,
                "size": null,
                "spin": false,
                "symbol": false,
                "title": "Vis dekningskart",
                "transform": null,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "Vis dekningskart",
          ],
          "type": "span",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
